networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java

```diff
@@ -87,18 +87,20 @@ public void disconnectImmediately(
   }
 
   @Override
-  public void disconnectCleanly(final DisconnectReason reason) {
+  public SafeFuture<?> disconnectCleanly(final DisconnectReason reason) {
     connected.set(false);
     disconnectReason = Optional.of(reason);
     disconnectLocallyInitiated = true;
-    disconnectRequestHandler
+    return disconnectRequestHandler
         .requestDisconnect(reason)
-        .finish(
-            () ->
-                disconnectImmediately(Optional.of(reason), true), // Request sent now close our side
+        .thenRun(
+            () -> // Request sent now close our side
+            disconnectImmediately(Optional.of(reason), true))
+        .exceptionally(
             error -> {
               LOG.debug("Failed to disconnect from " + getId() + " cleanly.", error);
               disconnectImmediately(Optional.of(reason), true);
+              return null;
             });
```

Comment:
Might be cleaner to use `handle` here:
```suggestion
        .handle(
            (__, error) -> {
              if (error != null) {
                LOG.debug("Failed to disconnect from " + getId() + " cleanly.", error);
              }
            disconnectImmediately(Optional.of(reason), true));
            return null;
         });
```

Reply:
Done.